docs/dev/client.rst

```diff
@@ -0,0 +1,79 @@
+
+The ``NubladoClient`` is designed to make interaction with JuptyerHub and Jupyterlab, as they are configured in the RSP environment, easy.
+
+A particular instance of the client represents a single user in a particular RSP environment.
+A user, in this context, means a token (which will have a set of scopes allowing various actions within the RSP) bound to a username.
+Both of these will be available in the service you are writing with each request, in the ``X-Auth-Request-Token`` and ``X-Auth-Request-User`` headers on the request.
+You should not, in general, need to go to Gafaelfawr to extract any further information about the token, but you must be prepared to handle 401s and 403s in case the token you have is invalid, expired, or does not grant sufficient scope for the service you want to use.
+
+Sequence of Events
+==================
+
+A typical interaction with the client usually looks like this:
+
+#. Authenticate to the Hub with the ``auth_to_hub()`` method.
+#. Determine whether or not you have a running lab with ``is_lab_stopped()``.
+#. If you need to, spawn a lab with ``spawn_lab()``.
+#. Wait for the lab to spawn by looping through ``watch_spawn_progress()`` until you get a progress message indicating the lab is ready.
+#. Authenticate to the Lab with ``auth_to_lab()``
+#. Use ``open_lab_session()`` as a context manager to interact with your lab.
+#. Do whatever it is you wanted to do with the lab; more below.
+#. When done, use ``stop_lab()`` to shut down the lab, if desired.
+
+The ``client_test`` test in the client test suite steps through this process and may be a useful model.
+
+.. _lab-interaction:
+
+Interacting With The Lab
+========================
+
+``NubladoClient`` provides three methods of interacting with a spawned lab.  These are methods on the ``JupyterLabSession`` object you will have available inside the session context manager.  They are:
+
+#. ``run_python()``.  This runs a string representing arbitrary Python code and returns streamed results (which is to say, ``stdout`` wihin each cell).  Of course, this code will run in the Lab environment specified by the session, which will be bound to a particuar kernel (usually ``LSST`` within the RSP), and will have access to whatever is installed in that kernel.
+#. ``run_notebook()``.  This runs each cell of a supplied notebook using ``run_python()``, accumulating the results and ultimately returning them as a list of cell outputs.
+#.  ``run_notebook_via_rsp_extension()``.  As is obvious from the name, this is not a standard Jupyter feature.  This runs a notebook in the same way ``Times-Square`` does, by using a Rubin Observatory-specific extension within the user lab that, in turn, uses ``nbconvert`` to execute notebooks and return their rendered form.  If you need to execute a notebook and capture output that did not go to stdout (for instance, the Javascript created by a Bokeh call, that will ultimately run in your browser), this is at present the way to do it.
```

Comment:
```suggestion
#.  ``run_notebook_via_rsp_extension()``.  This executes a notebook via the ``nbexec`` RSP JupyterLab extension.  `Times Square <https://times-square.lsst.io>`__ and `Noteburst <https://noteburst.lsst.io>`__ use this method. The ``nbexec`` JupyterLab extensions runs within the user lab, and in turn, uses ``nbconvert`` to execute notebooks and return their rendered form.  If you need to execute a notebook and capture output that did not go to stdout (for instance, the Javascript created by a Bokeh call, that will ultimately run in your browser), this is at present the way to do it.
```

Reply:
Noted (it's not `nbexec` anymore, just the `/rubin/execution` endpoint of the server extension, but I'll call it out).

Reply:
Done.